File: web/src/lib/player-stats.ts
// A player's numbers — the web twin of iOS `PlayerStats` and `PlayerGameLog`
// (StatSideShared/Models/PlayerStats.swift).
//
// **ESPN's categories and column labels are carried through, never named
// here** — the box score's lesson (2026-09-05) without amendment. A
// quarterback leads with `passing`, a receiver with `receiving`, the NBA
// ships `averages` and `totals`, and hockey splits `goaltender` from the
// skaters. ESPN orders a player's categories by what they actually do, so
// the first one is the player's own table — which is what the This season
// card reads.
//
// Every season the player has a line for is a row, with the club they
// played it for. That *is* the Career tab: no aggregation, and ESPN's own
// `totals` as the career line rather than a sum the app computed
// (2026-09-24).

export interface PlayerSeasonLine {
  /** ESPN's season year — the *ending* year for basketball and hockey. */
  year: number;
  /** "2026", "2025-26" — whatever ESPN calls it. */
  label: string;
  teamId?: string;
  /** The club's name, from the payload's own `teams` map. */
  teamName?: string;
  /** The club's abbreviation, from the same map — the Career row's caption. */
  teamAbbreviation?: string;
  position?: string;
  values: string[];
}

export interface PlayerStatsCategory {
  /** ESPN's machine name: "passing", "averages", "goaltender". */
  id: string;
  /** ESPN's heading: "Passing", "Regular Season Averages". */
  title: string;
  /** Column headers — "GP", "YDS", "TD". */
  labels: string[];
  /**
   * Stable column keys — "gamesPlayed", "passingYards". What the headline
   * registry matches on, so a relabelled column can't move which number
   * the card shows.
   */
  names: string[];
  /**
   * ESPN's full column names — "Passing Yards". Not its `descriptions[]`,
   * which are glossary sentences and read like a textbook in a list.
   */
  displayNames: string[];
  /** Oldest first, as ESPN sends them. */
  seasons: PlayerSeasonLine[];
  /** ESPN's career line, positionally paired with `labels`. Empty when it
   *  doesn't match the header. */
  career: string[];
}

export interface PlayerStats {
  categories: PlayerStatsCategory[];
}

/** One of the This season card's numbers. */
export interface PlayerHeadline {
  label: string;
  spokenLabel: string;
  value: string;
}

/** A player traded mid-season has two lines for one year. */
export function seasonLineId(line: PlayerSeasonLine): string {
  return `${line.year}-${line.teamId ?? ""}`;
}

/** The value in the column ESPN names `name`, for one line. */
export function categoryValue(
  category: PlayerStatsCategory,
  name: string,
  values: string[]
): string | undefined {
  const index = category.names.indexOf(name);
  if (index === -1 || index >= values.length) return undefined;
  const value = values[index];
  return value === "" || value === "-" ? undefined : value;
}

/** The lines for one ESPN season year — usually one, two for a player
 *  traded mid-season. */
export function linesForYear(
  category: PlayerStatsCategory,
  year: number
): PlayerSeasonLine[] {
  return category.seasons.filter((line) => line.year === year);
}

const SKATER_CATEGORIES = new Set([
  "center",
  "leftWing",
  "rightWing",
  "defense",
  "forward",
  "skater",
]);

/**
 * Which columns lead, keyed by ESPN's category name rather than by league —
 * a receiver in either football league is the same question. A category
 * this doesn't know shows its first three columns after games played, which
 * is ESPN's own order of importance.
 *
 * A straight copy of iOS `PlayerStats.headlineNames(for:)`; the two must
 * agree or one player reads differently on the two platforms.
 */
export function headlineNames(category: PlayerStatsCategory): string[] {
  switch (category.id) {
    case "passing":
      return ["passingYards", "passingTouchdowns", "interceptions"];
    case "rushing":
      return ["rushingYards", "rushingTouchdowns", "yardsPerRushAttempt"];
    case "receiving":
      return ["receptions", "receivingYards", "receivingTouchdowns"];
    case "averages":
      return ["avgPoints", "avgRebounds", "avgAssists"];
    case "goaltender":
      return ["wins", "avgGoalsAgainst", "savePct"];
    default:
      if (SKATER_CATEGORIES.has(category.id)) {
        return ["goals", "assists", "points"];
      }
      return category.names
        .filter((name) => name !== "gamesPlayed" && name !== "games")
        .slice(0, 3);
  }
}

/**
 * The This season card's contents: games played, then the two or three
 * numbers the player's own category is about — or nothing, which hides the
 * card. A player with no line this season has no current season, and a row
 * of zeroes would say otherwise.
 *
 * A traded player's lines for the year are not summed: the last line ESPN
 * lists stands, because the app does not publish a figure it computed.
 */
export function seasonHeadlines(
  stats: PlayerStats,
  espnYear: number
): PlayerHeadline[] {
  const category = stats.categories[0];
  if (!category) return [];
  const line = linesForYear(category, espnYear).at(-1);
  if (!line) return [];

  const wanted = ["gamesPlayed", "games", ...headlineNames(category)];
  const seen = new Set<string>();
  const headlines: PlayerHeadline[] = [];
  for (const name of wanted) {
    const index = category.names.indexOf(name);
    if (index === -1) continue;
    const value = categoryValue(category, name, line.values);
    const label = category.labels[index];
    if (value === undefined || label === undefined || seen.has(label)) continue;
    seen.add(label);
    headlines.push({
      label,
      spokenLabel: category.displayNames[index] ?? label,
      value,
    });
  }
  return headlines;
}

/** The season label the This season card names, or undefined. */
export function seasonLabelFor(
  stats: PlayerStats,
  espnYear: number
): string | undefined {
  const category = stats.categories[0];
  return category ? linesForYear(category, espnYear).at(-1)?.label : undefined;
}

// --- Game log -----------------------------------------------------------

export interface PlayerGameLogEntry {
  eventId: string;
  /** ISO string, as ESPN sent it — survives the server/client boundary. */
  date?: string;
  /** Football only; ESPN sends none for basketball or hockey. */
  week?: number;
  /** True for "@", false for "vs". */
  isAway: boolean;
  opponentId?: string;
  opponentAbbreviation?: string;
  opponentName?: string;
  opponentLogoUrl?: string;
  /** The player's own side that night — a traded player's log crosses
   *  clubs, so this is per game, not per page. */
  teamId?: string;
  /** "W", "L", "T" — or undefined for a game not yet decided. */
  result?: string;
  teamScore?: string;
  opponentScore?: string;
  /** "West Semifinals - Game 4". */
  note?: string;
  /** Positionally paired with the log's `labels`. */
  values: string[];
}

/** "2025-26 Regular Season", "2025-26 Postseason". */
export interface PlayerGameLogSection {
  title: string;
  /** Newest first, the way ESPN lists them. */
  entries: PlayerGameLogEntry[];
}

export interface PlayerGameLog {
  labels: string[];
  names: string[];
  /** Newest season type first. */
  sections: PlayerGameLogSection[];
  /** The seasons ESPN will answer for, newest first, in ESPN's numbering. */
  availableSeasons: number[];
  /** The season this log is for, in ESPN's numbering. */
  season?: number;
}

export function gameLogIsEmpty(log: PlayerGameLog): boolean {
  return log.sections.every((section) => section.entries.length === 0);
}

/**
 * The short line a game row prints — the same columns the This season card
 * leads with, looked up by name in this log's own header. iOS
 * `PlayerGameLog.headline(for:category:)`.
 */
export function gameLogHeadline(
  log: PlayerGameLog,
  entry: PlayerGameLogEntry,
  category: string | undefined
): string {
  let picks: string[];
  switch (category) {
    case "passing":
      picks = ["passingYards", "passingTouchdowns", "interceptions"];
      break;
    case "rushing":
      picks = ["rushingYards", "rushingTouchdowns"];
      break;
    case "receiving":
      picks = ["receptions", "receivingYards", "receivingTouchdowns"];
      break;
    case "averages":
    case "totals":
      picks = ["points", "totalRebounds", "assists"];
      break;
    case "goaltender":
      picks = ["saves", "savePct", "goalsAgainst"];
      break;
    default:
      picks =
        category !== undefined && SKATER_CATEGORIES.has(category)
          ? ["goals", "assists", "points"]
          : [];
  }

  const parts: string[] = [];
  for (const name of picks) {
    const index = log.names.indexOf(name);
    if (index === -1 || index >= entry.values.length || index >= log.labels.length) {
      continue;
    }
    parts.push(`${entry.values[index]} ${log.labels[index]}`);
  }
  if (parts.length === 0) {
    // Nothing named — the first three columns, ESPN's own order of
    // importance for this log.
    for (let index = 0; index < Math.min(3, log.labels.length, entry.values.length); index += 1) {
      parts.push(`${entry.values[index]} ${log.labels[index]}`);
    }
  }
  return parts.join(" · ");
}

/** "W 33-30", the player's side first. */
export function gameLogResult(entry: PlayerGameLogEntry): string | undefined {
  if (!entry.result) return undefined;
  if (entry.teamScore === undefined || entry.opponentScore === undefined) {
    return entry.result;
  }
  return `${entry.result} ${entry.teamScore}-${entry.opponentScore}`;
}

File: StatSideShared/Models/PlayerStats.swift
import Foundation

/// A player's season-by-season numbers, as ESPN tables them.
///
/// **ESPN's categories and column labels are carried through, never named
/// here** — the box score's lesson (2026-09-05) without amendment. The
/// categories are per league and per position: a quarterback leads with
/// `passing`, a receiver with `receiving`, the NBA ships `averages` and
/// `totals`, and hockey splits `goaltender` from the skaters. ESPN orders a
/// player's categories by what they actually do, so the first one is the
/// player's own table — which is what the Current season card reads.
///
/// Every season the player has a line for is a row, with the club they
/// played it for. That *is* the Career tab: no aggregation, and ESPN's own
/// `totals` as the career line rather than a sum the app computed
/// (2026-09-24, answering E20's Career row).
nonisolated struct PlayerStats: Sendable, Hashable {
    struct Category: Sendable, Hashable, Identifiable {
        /// ESPN's machine name: "passing", "averages", "goaltender".
        let id: String
        /// ESPN's heading: "Passing", "Regular Season Averages".
        let title: String
        /// Column headers — "GP", "YDS", "TD".
        let labels: [String]
        /// Stable column keys — "gamesPlayed", "passingYards". What the
        /// headline registry matches on, so a relabelled column can't move
        /// which number the card shows.
        let names: [String]
        /// ESPN's full column names — "Passing Yards", "Points Per Game".
        /// Not its `descriptions[]`, which are glossary sentences ("The
        /// total passing yards.") and read like a textbook in a list.
        let displayNames: [String]
        /// Oldest first, as ESPN sends them.
        let seasons: [SeasonLine]
        /// ESPN's career line, positionally paired with `labels`. Empty when
        /// it doesn't match the header.
        let career: [String]
    }

    struct SeasonLine: Sendable, Hashable, Identifiable {
        /// ESPN's season year — the *ending* year for basketball and hockey.
        let year: Int
        /// "2026", "2025-26", "25-26" — whatever ESPN calls it.
        let label: String
        let teamId: String?
        /// The club's name, from the payload's own `teams` map.
        let teamName: String?
        let position: String?
        let values: [String]

        /// A player traded mid-season has two lines for one year.
        var id: String { "\(year)-\(teamId ?? "")" }
    }

    let categories: [Category]

    /// The categories with at least one season line — what the Stats and
    /// Career tabs draw. A category ESPN names with no rows would otherwise
    /// be a table of headers with nothing under them. The web's
    /// `categoriesWithLines` is the same rule.
    var categoriesWithLines: [Category] { categories.filter { !$0.seasons.isEmpty } }

    static let empty = PlayerStats(categories: [])
}

nonisolated extension PlayerStats.Category {
    /// The value in the column ESPN names `name`, for one line.
    func value(_ name: String, in values: [String]) -> String? {
        guard let index = names.firstIndex(of: name), values.indices.contains(index) else { return nil }
        let value = values[index]
        return value.isEmpty || value == "-" ? nil : value
    }

    /// The lines for one ESPN season year — usually one, two for a player
    /// traded mid-season.
    func lines(for year: Int) -> [PlayerStats.SeasonLine] {
        seasons.filter { $0.year == year }
    }
}

// MARK: - Headline numbers

nonisolated extension PlayerStats {
    /// One of the Current season card's numbers.
    struct Headline: Sendable, Hashable, Identifiable {
        let label: String
        let spokenLabel: String
        let value: String
        var id: String { label }
    }

    /// The Current season card's contents: games played, then the two or
    /// three numbers the player's own category is about — or nothing, which
    /// hides the card. A player with no line this season has no current
    /// season, and a column of zeroes would say otherwise.
    ///
    /// A traded player's lines for the year are not summed: the first line
    /// ESPN lists stands, because the app does not publish a figure it
    /// computed (E20's Career rule, applied here too).
    func headlines(forSeason year: Int) -> [Headline] {
        guard let category = categories.first,
              let line = category.lines(for: year).last else { return [] }
        let wanted = ["gamesPlayed", "games"] + Self.headlineNames(for: category)
        var seen: Set<String> = []
        return wanted.compactMap { name -> Headline? in
            guard let index = category.names.firstIndex(of: name),
                  let value = category.value(name, in: line.values),
                  seen.insert(category.labels[index]).inserted
            else { return nil }
            let spoken = category.displayNames.indices.contains(index)
                ? category.displayNames[index] : category.labels[index]
            return Headline(label: category.labels[index], spokenLabel: spoken, value: value)
        }
    }

    /// Which columns lead, keyed by ESPN's category name rather than by
    /// league — a receiver in either football league is the same question.
    /// A category this doesn't know shows its first three columns after
    /// games played, which is ESPN's own order of importance.
    static func headlineNames(for category: Category) -> [String] {
        switch category.id {
        case "passing": ["passingYards", "passingTouchdowns", "interceptions"]
        case "rushing": ["rushingYards", "rushingTouchdowns", "yardsPerRushAttempt"]
        case "receiving": ["receptions", "receivingYards", "receivingTouchdowns"]
        case "averages": ["avgPoints", "avgRebounds", "avgAssists"]
        case "goaltender": ["wins", "avgGoalsAgainst", "savePct"]
        case "center", "leftWing", "rightWing", "defense", "forward", "skater":
            ["goals", "assists", "points"]
        default:
            Array(category.names.filter { $0 != "gamesPlayed" && $0 != "games" }.prefix(3))
        }
    }
}

// MARK: - Game log

/// One season of a player's games, as ESPN logs them.
nonisolated struct PlayerGameLog: Sendable, Hashable {
    /// A run of columns that belong together — "Passing" over eleven,
    /// "Rushing" over five. Football's log is one flat row with repeated
    /// labels (two "YDS", two "TD"), and the groups are what tells them
    /// apart.
    struct ColumnGroup: Sendable, Hashable {
        let title: String
        let count: Int
    }

    struct Entry: Sendable, Hashable, Identifiable {
        let eventId: String
        let date: Date?
        /// Football only; ESPN sends none for basketball or hockey.
        let week: Int?
        /// True for "@", false for "vs".
        let isAway: Bool
        let opponentId: String?
        let opponentAbbreviation: String?
        let opponentName: String?
        let opponentLogoURL: URL?
        /// The player's own side that night — a traded player's log crosses
        /// clubs, so this is per game, not per page.
        let teamId: String?
        let teamAbbreviation: String?
        let teamLogoURL: URL?
        /// "W", "L", "T" — or nil for a game not yet decided.
        let result: String?
        let teamScore: String?
        let opponentScore: String?
        /// "West Semifinals - Game 4", "CFP Semifinal at the …".
        let note: String?
        /// Positionally paired with the log's `labels`.
        let values: [String]
        var id: String { eventId }
    }

    /// "2025-26 Regular Season", "2025-26 Postseason".
    struct Section: Sendable, Hashable, Identifiable {
        let title: String
        /// Newest first, the way ESPN lists them.
        let entries: [Entry]
        var id: String { title }
    }

    let labels: [String]
    let names: [String]
    let groups: [ColumnGroup]
    /// Newest season type first.
    let sections: [Section]
    /// The seasons ESPN will answer for, newest first — the season menu.
    let availableSeasons: [Int]
    /// The season this log is for.
    let season: Int?

    static let empty = PlayerGameLog(labels: [], names: [], groups: [], sections: [],
                                     availableSeasons: [], season: nil)

    var isEmpty: Bool { sections.allSatisfy(\.entries.isEmpty) }
}

nonisolated extension PlayerGameLog {
    /// The short line a game row prints — the same columns the Current
    /// season card leads with, looked up by name in this log's own header.
    func headline(for entry: Entry, category: String?) -> String {
        let picks: [String] = switch category {
        case "passing": ["passingYards", "passingTouchdowns", "interceptions"]
        case "rushing": ["rushingYards", "rushingTouchdowns"]
        case "receiving": ["receptions", "receivingYards", "receivingTouchdowns"]
        case "averages", "totals": ["points", "totalRebounds", "assists"]
        case "goaltender": ["saves", "savePct", "goalsAgainst"]
        case "center", "leftWing", "rightWing", "defense", "forward", "skater":
            ["goals", "assists", "points"]
        default: []
        }
        var parts: [String] = []
        for name in picks {
            guard let index = names.firstIndex(of: name),
                  entry.values.indices.contains(index), labels.indices.contains(index)
            else { continue }
            parts.append("\(entry.values[index]) \(labels[index])")
        }
        if parts.isEmpty {
            // Nothing named — the first three columns, which is ESPN's own
            // order of importance for this log.
            parts = zip(labels, entry.values).prefix(3).map { "\($1) \($0)" }
        }
        return parts.joined(separator: " · ")
    }
}
